docusaurus/docs/create-a-plugin/develop-a-plugin/manage-npm-dependencies.md

```diff
@@ -0,0 +1,52 @@
+---
+id: manage-npm-dependencies
+title: Manage NPM dependencies
+description: How to manage frontend NPM dependencies in Grafana plugins.
+keywords:
+  - grafana
+  - plugins
+  - plugin
+  - github
+  - npm
+  - dependencies
+sidebar_position: 5
+---
+
+# Manage frontend NPM dependencies in a Grafana plugin
+
+Frontend plugins in Grafana have their own unique dependencies, as well as dependencies that are shared with the Grafana application during runtime. This document focuses on how these shared dependencies, particularly the `@grafana` npm packages, are handled.
+
+It's important to understand that while a plugin specifies the expected versions of these dependencies in its `package.json` file, they are dynamically linked to the Grafana version at runtime.
+
+# Dynamic dependency linking
+
+When you develop your plugin, its `package.json` may reference a specific version of a `@grafana` npm package, such as `@grafana/ui: 9.5.1`.
+
+However, when the plugin is installed and executed within a Grafana instance, it inherits the version of the `@grafana` packages that the Grafana application is using. For example, if the Grafana version is 10.0.0, then the plugin uses version 10.0.0 of the shared dependencies from the Grafana application.
+
+# Dependency sharing mechanism
+
+To facilitate this dynamic dependency linking, Grafana employs SystemJS for loading frontend plugin code and sharing some of the Grafana application's npm dependencies with plugins.
+
+Grafana makes the decision to share dependencies for one of two reasons:
+
+- **Singleton dependency requirement:** In some cases, only a single instance of a dependency can exist during runtime.
+- **Performance optimization:** Sharing dependencies can enhance performance, especially when dealing with large dependency codebases.
+
+# Requirements for sharing dependencies
+
+To successfully share dependencies, you must define two key components:
+
+- **SystemJS import map in Grafana:** The dependency must be listed in a SystemJS import map in the Grafana application.
+- **Plugin build tool configuration:** The dependency must be externalized in the plugin's build tool configuration, which is primarily done using Webpack.
+
```

Comment:
I'm wondering if we should either remove this section entirely or add a note at the bottom explaining that attempting to change the externalised dependencies in build configs is not supported and will likely lead to plugin load failures?

```suggestion
# Requirements for sharing dependencies

To successfully share dependencies, you must define two key components:

- **SystemJS import map in Grafana:** The dependency must be listed in a SystemJS import map in the Grafana application.
- **Plugin build tool configuration:** The dependency must be externalized in the plugin's build tool configuration, which is primarily done using Webpack.

:::caution

Customising the build tool configuration to change the external dependencies is not supported and will likely lead to plugin loading failure or bugs.

:::

```

Reply:
@jackw yeah I'm struggling with this section, what should a plugin developer do with this information (if they can do anything...). Statements like "you must define" imply action, but these don't seem to be elements the dev would interact with?

Reply:
Yep, it's informative so nothing they can do. That's also why I'm questioning removing it.

Reply:
What if we say "Grafana defines" and remove "successfully," instead of "you must define" above?

Reply:
Let's remove "you must define" with "Grafana defines." If I understand the issue correctly, this should remove potential confusion.